mutation_reader.cc

```diff
@@ -233,26 +237,29 @@ class mutation_reader_merger {
     future<> fast_forward_to(position_range pr, db::timeout_clock::time_point timeout);
 };
 
-// Combines multiple mutation_readers into one.
-class combined_mutation_reader : public flat_mutation_reader::impl {
-    mutation_fragment_merger<mutation_reader_merger> _producer;
+/* Merge a non-decreasing stream of mutation fragments produced
+ * by a FragmentProducer into a strictly increasing stream.
```

Comment:
This is what the comment above `mutation_fragment_merger` says:
>  Merge a non-decreasing stream of mutation-fragments into strictly increasing stream.

Perhaps the author of that comment has something else in mind when they wrote "strictly increasing stream". E.g. maybe they understood two different range tombstones but with the same position to have a well defined order w.r.t each other, or whatever.

But yeah, it's confusing. I guess I'll have to fix both comments.

Reply:
Maybe that author of that code didn't know what he was talking about. I know him well, and he often does that.

Reply:
(☞ﾟヮﾟ)☞
Heh, I tend to do that too.

Reply:
(he learned about the non-striclty part much later and through the hard way)

Reply:
Well the stream is mostly strictly increasing, range tombstones are the only exception (are they?)

And the reader's purpose is indeed merging fragments with the same position. That's basically all it does. With the small exception of range tombstones.

I guess it's best to write exactly what it does in the comment. That is, given a stream of fragment batches, where every batch has the same position but different batches have different positions, it returns a non-decreasing stream of fragments, where two fragments have the same position *only* if they are range tombstones, and otherwise have different positions.

Reply:
I think that's over doing it. :)) "Merges multiple fragment streams into a single one." is enough. No need to explain here what a valid fragment stream is. Also, the fact that partitions with the same key and rows with the same key... etc. are merged is a fundamental property of our data model.

Reply:
> The resulting stream will not be strictly increasing, it monotonic, but not strictly monotonic. Range tombstones are allowed to have the same position as other range tombstones.

We *must* fix it, it's a constant thorn.

Reply:
> We must fix it, it's a constant thorn.

I guess we could compare range tombstones lexicographically by the (left-end, right-end) pair, with some added weights to account for closed/open interval bounds. Would complicate `position_in_partition` even more but meh.

Then two range tombstones would be mergeable if they had both ends and their weights equal.

Reply:
We already take bound weight into consideration, when comparing `position_in_partition`. By bound weight I mean before/after key, a range tombstone cannot represent a closed interval bound.

To fix this we'd need deoverlapping of range tombstones, then we'd guarantee to have each given point in the clustering domain covered by exactly one range tombstone.

Reply:
Deoverlapping would be nice, but we don't really *need* to do it to solve this particular problem of non-strictly increasing fragments. The least effort solution would be to ensure that different range tombstones have different `position_in_partition`s (even if they have the same left-end). Hence the suggestion to take the right-end into account too (so the position of a range tombstone would depend on *both* the left and the right end).

Reply:
I don't see how it can be done, how can we cram two positions into a single one?

Reply:
I mean it *can* be done, but it feels like a dirty hack.

Reply:
It's just modifying the definition of "position". The result would be less intuitive and probably not reflect a "position" at all any more, but it would have the desirable mathematical properties (two fragments are different => they have different positions).

I agree it's a hack.

Reply:
Also, I think it is technically allowed for two range tombstones to have identical start and end bounds, with only the timestamp being different.

Reply:
I might be wrong on this one (but would bet that I'm right).

Reply:
If we do the work, we may as well deoverlap. I don't think it's harder.

Reply:
Most components already use `range_tombstone_list` to collect and emit range tombstones. If we add deoverlapping support there we already cover many users. Then we'd need to make the combined reader use this too, as well as sstable readers. Memtables and cache could deoverlap on insert.

Reply:
Created https://github.com/scylladb/scylla/issues/7687 so this discussion is not lost